### ISSUE TITLE
Support for "IP Transparency" in envoy for ingress gateways

### ISSUE THREAD

**Issue by apollo13** [hashicorp/consul#8665]
#### Feature Description

It would be great if the ingress gateway configuration supported configuring IP transparency ( https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/other_features/ip_transparency )

#### Use Case(s)

Even if services are inside a service mesh, IP based ACLs can make sense. It is understandable if the ingress gateway doesn't support fancy things like authentication etc, but passing on the real source IP would be helpful for many people I guess.

**Comment by blake** [comment]
Hi @apollo13,

Could you provide a bit more info about your use case for this feature? Are you looking for an upstream proxy to evaluate the downstream client IP and take some action based on an allow list, or are you simply looking to pass the client IP information to the upstream application?

If the latter, this should be possible as of Consul 1.11.0 by configuring a service router to add the client IP to the upstream request using the [HTTP header modification feature](https://www.consul.io/docs/connect/config-entries/service-router#httpheadermodifiers).

```hcl
# web-service-router.hcl
Kind = "service-router"
Name = "web"
Routes = [
  {
    Match {
      HTTP {
        PathPrefix = "/"
      }
    }

    Destination {
      RequestHeaders {
        Add {
          X-Downstream-Remote-Address = "%DOWNSTREAM_REMOTE_ADDRESS%"
        }
      }
    }
  }
]
```

The upstream service will then receive the client IP in the configured HTTP header.

```
GET / HTTP/1.1 
host: localhost:9191 
user-agent: curl/7.64.1 
accept: */* 
x-forwarded-proto: http 
x-request-id: 9d516094-ca1e-4c75-be47-473d322ce797 
x-downstream-remote-address: 192.0.2.10:62030 
x-envoy-expected-rq-timeout-ms: 15000 
```

**Comment by apollo13** [comment]
> Are you looking for an upstream proxy to evaluate the downstream client IP and take some action based on an allow list

Yes, I know that I can add headers etc but it is kinda to add that to services that expect something else. That said even with transparency it probably would need a transparent proxy at the end to keep the back route consistent.

> If the latter, this should be possible as of Consul 1.11.0 by configuring a service router to add the client IP to the upstream request using the HTTP header modification feature.

That is a L7 feature and doesn't work in TCP mode right? I still have to use tcp mode due to #8283

**Comment by blake** [comment]
> That is a L7 feature and doesn't work in TCP mode right? I still have to use tcp mode due to #8283

Yes, this is an L7 feature that only works with HTTP.

From the remaining transparency modes, which one best addresses your requirements?

* Proxy protocol
* Original source listener filter
* Original source HTTP filter

**Comment by apollo13** [comment]
In an ideal world Original source listener filter would be the preferred since it requires no application changes. But this requires iptables trickery and I am not sure how feasible this would be to get into nomad (although it could probably be done inside the connect proxy itself). So the most realistic thing is using proxy protocol to connect to the backend service, which is luckily supported by many tools.

**Comment by Legogris** [comment]
Just chiming in that proxy protocol seems like a good way to handle this
